File: cpp/dijkstra/1238.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#define INF 1e9

using namespace std;
vector<pair<int,int>> graph[1001];
int d[1001];
int max_elem = 0;

void dijkstra(int start){
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
    pq.push({0,start});
    d[start] = 0;
    while(!pq.empty()){
        int dist = pq.top().first;
        int now = pq.top().second;
        pq.pop();

        if(d[now] < dist){
            continue;
        }

        for(int i=0;i<graph[now].size();i++){
            int cost = dist + graph[now][i].second;
            if(cost >= d[graph[now][i].first]){
                continue;
            }
            pq.push({cost, graph[now][i].first});
            d[graph[now][i].first] = cost;
        }
    }
}

int main() {
    int n,m,x;
    cin >> n >> m >> x;
    for(int i=0;i<m;i++){
        int a,b,c;
        cin >> a >> b >> c;
        graph[a].push_back({b,c});
    }

    for(int i=1;i<=n;i++){
        int value = 0;
        fill(d,d+1001,INF);
        dijkstra(i);
        value += d[x];
        fill(d,d+1001,INF);
        dijkstra(x);
        value += d[i];
        max_elem = max(max_elem, value);
    }

    cout << max_elem << '\n';
}

File: python/boj/shortest_path/1238.py
import sys
import heapq
input = sys.stdin.readline
INF = 1e9

n,m,x = map(int,input().split())

graph = [[] for _ in range(n+1)]

for _ in range(m):
    a,b,c, = map(int,input().split())
    graph[a].append((b,c))

def dijkstra(start,d):
    q = []
    heapq.heappush(q,(0,start))
    d[start] = 0
    while q:
        dist, now = heapq.heappop(q)
        if d[now] < dist:
            continue
        for i in graph[now]:
            cost = dist + i[1]
            if cost >= d[i[0]]:
                continue
            d[i[0]] = cost
            heapq.heappush(q, (cost, i[0]))

result = 0

for i in range(1,n+1):
    total = 0
    d = [INF] * (n+1)
    dijkstra(i,d)
    total += d[x]
    new_d = [INF] * (n+1)
    dijkstra(x,new_d)
    total += new_d[i]
    result = max(result, total)


print(result)

    



